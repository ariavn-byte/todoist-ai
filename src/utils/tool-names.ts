/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
