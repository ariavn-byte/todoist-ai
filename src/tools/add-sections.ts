import type { Section } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { ADD_TASKS, FIND_TASKS, GET_OVERVIEW, FIND_SECTIONS } = ToolNames

const SectionSchema = z.object({
    name: z.string().min(1).describe('The name of the section.'),
    projectId: z.string().min(1).describe('The ID of the project to add the section to.'),
})

const ArgsSchema = {
    sections: z.array(SectionSchema).min(1).describe('The array of sections to add.'),
}

const addSections = {
    name: ToolNames.ADD_SECTIONS,
    description: 'Add one or more new sections to projects.',
    parameters: ArgsSchema,
    async execute({ sections }, client) {
        const newSections = await Promise.all(sections.map((section) => client.addSection(section)))
        const textContent = generateTextContent({ sections: newSections })

        return getToolOutput({
            textContent,
            structuredContent: {
                sections: newSections,
                totalCount: newSections.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    sections,
}: {
    sections: Section[]
}) {
    const count = sections.length
    const sectionList = sections
        .map((section) => `â€¢ ${section.name} (id=${section.id}, projectId=${section.projectId})`)
        .join('\n')

    const summary = `Added ${count} section${count === 1 ? '' : 's'}:\n${sectionList}`

    // Context-aware next steps for new sections
    const nextSteps: string[] = []

    if (count === 1) {
        const section = sections[0]
        if (section) {
            nextSteps.push(`Use ${ADD_TASKS} with sectionId=${section.id} to add your first tasks`)
            nextSteps.push(`Use ${FIND_TASKS} with sectionId=${section.id} to verify setup`)
            nextSteps.push(
                `Use ${GET_OVERVIEW} with projectId=${section.projectId} to see project organization`,
            )
        }
    } else {
        // Group sections by project for better guidance
        const projectIds = [...new Set(sections.map((s) => s.projectId))]

        nextSteps.push(`Use ${ADD_TASKS} to add tasks to these new sections`)

        if (projectIds.length === 1) {
            nextSteps.push(
                `Use ${GET_OVERVIEW} with projectId=${projectIds[0]} to see updated project structure`,
            )
            nextSteps.push(
                `Use ${FIND_SECTIONS} with projectId=${projectIds[0]} to see all sections`,
            )
        } else {
            nextSteps.push(`Use ${GET_OVERVIEW} to see updated project structures`)
            nextSteps.push(`Use ${FIND_SECTIONS} to see sections in specific projects`)
        }
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { addSections }
