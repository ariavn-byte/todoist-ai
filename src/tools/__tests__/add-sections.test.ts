import type { Section, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    TEST_IDS,
    createMockSection,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { addSections } from '../add-sections.js'

// Mock the Todoist API
const mockTodoistApi = {
    addSection: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { ADD_TASKS, ADD_SECTIONS, FIND_TASKS } = ToolNames

describe(`${ADD_SECTIONS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('creating a single section', () => {
        it('should create a section and return mapped result', async () => {
            const mockApiResponse = createMockSection({
                id: TEST_IDS.SECTION_1,
                projectId: TEST_IDS.PROJECT_TEST,
                name: 'test-abc123def456-section',
            })

            mockTodoistApi.addSection.mockResolvedValue(mockApiResponse)

            const result = await addSections.execute(
                {
                    sections: [
                        { name: 'test-abc123def456-section', projectId: TEST_IDS.PROJECT_TEST },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly
            expect(mockTodoistApi.addSection).toHaveBeenCalledWith({
                name: 'test-abc123def456-section',
                projectId: TEST_IDS.PROJECT_TEST,
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 1 section:')
            expect(textContent).toContain('test-abc123def456-section')
            expect(textContent).toContain(`id=${TEST_IDS.SECTION_1}`)
            expect(textContent).toContain(`Use ${ADD_TASKS} with sectionId`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    sections: [
                        expect.objectContaining({
                            id: TEST_IDS.SECTION_1,
                            name: 'test-abc123def456-section',
                        }),
                    ],
                    totalCount: 1,
                }),
            )
        })

        it('should handle different section properties from API', async () => {
            const mockApiResponse = createMockSection({
                id: TEST_IDS.SECTION_2,
                projectId: 'project-789',
                sectionOrder: 2,
                name: 'My Section Name',
            })

            mockTodoistApi.addSection.mockResolvedValue(mockApiResponse)

            const result = await addSections.execute(
                { sections: [{ name: 'My Section Name', projectId: 'project-789' }] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addSection).toHaveBeenCalledWith({
                name: 'My Section Name',
                projectId: 'project-789',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 1 section:')
            expect(textContent).toContain('My Section Name')
            expect(textContent).toContain(`id=${TEST_IDS.SECTION_2}`)
            expect(textContent).toContain(`Use ${FIND_TASKS} with sectionId`)
        })
    })

    describe('creating multiple sections', () => {
        it('should create multiple sections and return mapped results', async () => {
            const mockSections = [
                createMockSection({
                    id: 'section-1',
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'First Section',
                }),
                createMockSection({
                    id: 'section-2',
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'Second Section',
                }),
                createMockSection({
                    id: 'section-3',
                    projectId: 'different-project',
                    name: 'Third Section',
                }),
            ]

            const [section1, section2, section3] = mockSections as [Section, Section, Section]
            mockTodoistApi.addSection
                .mockResolvedValueOnce(section1)
                .mockResolvedValueOnce(section2)
                .mockResolvedValueOnce(section3)

            const result = await addSections.execute(
                {
                    sections: [
                        { name: 'First Section', projectId: TEST_IDS.PROJECT_TEST },
                        { name: 'Second Section', projectId: TEST_IDS.PROJECT_TEST },
                        { name: 'Third Section', projectId: 'different-project' },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each section
            expect(mockTodoistApi.addSection).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.addSection).toHaveBeenNthCalledWith(1, {
                name: 'First Section',
                projectId: TEST_IDS.PROJECT_TEST,
            })
            expect(mockTodoistApi.addSection).toHaveBeenNthCalledWith(2, {
                name: 'Second Section',
                projectId: TEST_IDS.PROJECT_TEST,
            })
            expect(mockTodoistApi.addSection).toHaveBeenNthCalledWith(3, {
                name: 'Third Section',
                projectId: 'different-project',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 3 sections:')
            expect(textContent).toContain('First Section (id=section-1, projectId=')
            expect(textContent).toContain('Second Section (id=section-2, projectId=')
            expect(textContent).toContain(
                'Third Section (id=section-3, projectId=different-project)',
            )

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    sections: mockSections,
                    totalCount: 3,
                }),
            )
        })

        it('should handle sections for the same project', async () => {
            const mockSections = [
                createMockSection({
                    id: 'section-1',
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'To Do',
                }),
                createMockSection({
                    id: 'section-2',
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'In Progress',
                }),
            ]

            const [section1, section2] = mockSections as [Section, Section]
            mockTodoistApi.addSection
                .mockResolvedValueOnce(section1)
                .mockResolvedValueOnce(section2)

            const result = await addSections.execute(
                {
                    sections: [
                        { name: 'To Do', projectId: TEST_IDS.PROJECT_TEST },
                        { name: 'In Progress', projectId: TEST_IDS.PROJECT_TEST },
                    ],
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 2 sections:')
            expect(textContent).toContain(`Use ${ADD_TASKS} to add tasks to these new sections`)
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Section name is required')
            mockTodoistApi.addSection.mockRejectedValue(apiError)

            await expect(
                addSections.execute(
                    { sections: [{ name: '', projectId: TEST_IDS.PROJECT_TEST }] },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Section name is required')
        })

        it('should handle partial failures in multiple sections', async () => {
            const mockSection = createMockSection({
                id: 'section-1',
                projectId: TEST_IDS.PROJECT_TEST,
                name: 'First Section',
            })

            mockTodoistApi.addSection
                .mockResolvedValueOnce(mockSection)
                .mockRejectedValueOnce(new Error('API Error: Invalid project ID'))

            await expect(
                addSections.execute(
                    {
                        sections: [
                            { name: 'First Section', projectId: TEST_IDS.PROJECT_TEST },
                            { name: 'Second Section', projectId: 'invalid-project' },
                        ],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Invalid project ID')
        })
    })
})
