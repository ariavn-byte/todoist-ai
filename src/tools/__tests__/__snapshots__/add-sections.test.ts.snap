// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`add-sections tool creating a single section should create a section and return mapped result 1`] = `
"Added 1 section:
• test-abc123def456-section (id=section-123, projectId=6cfCcrrCFg2xP94Q)
Next:
- Use add-tasks with sectionId=section-123 to add your first tasks
- Use find-tasks with sectionId=section-123 to verify setup
- Use get-overview with projectId=6cfCcrrCFg2xP94Q to see project organization"
`;

exports[`add-sections tool creating a single section should handle different section properties from API 1`] = `
"Added 1 section:
• My Section Name (id=section-456, projectId=project-789)
Next:
- Use add-tasks with sectionId=section-456 to add your first tasks
- Use find-tasks with sectionId=section-456 to verify setup
- Use get-overview with projectId=project-789 to see project organization"
`;

exports[`add-sections tool creating multiple sections should create multiple sections and return mapped results 1`] = `
"Added 3 sections:
• First Section (id=section-1, projectId=6cfCcrrCFg2xP94Q)
• Second Section (id=section-2, projectId=6cfCcrrCFg2xP94Q)
• Third Section (id=section-3, projectId=different-project)
Next:
- Use add-tasks to add tasks to these new sections
- Use get-overview to see updated project structures
- Use find-sections to see sections in specific projects"
`;

exports[`add-sections tool creating multiple sections should handle sections for the same project 1`] = `
"Added 2 sections:
• To Do (id=section-1, projectId=6cfCcrrCFg2xP94Q)
• In Progress (id=section-2, projectId=6cfCcrrCFg2xP94Q)
Next:
- Use add-tasks to add tasks to these new sections
- Use get-overview with projectId=6cfCcrrCFg2xP94Q to see updated project structure
- Use find-sections with projectId=6cfCcrrCFg2xP94Q to see all sections"
`;
